### PR TITLE
Fix #1878 (SA1133CodeFixProvider crash with System.InvalidCastException)

### LINKED ISSUE #1878: SA1133CodeFixProvider crash
Given the following code :

```csharp
namespace Stylecop_rc1_bug_repro
{
    using System;

    internal class Program
    {
        [Foo, Bar]
        private static void Main(string[] args)
        {
        }
    }

    internal class FooAttribute : Attribute
    {
    }

    internal class BarAttribute : Attribute
    {
    }
}
```

SA1133 is raised on `Bar` usage but activating the CodeFix throw an exception :

```text
Exception thrown: 'System.InvalidCastException' in StyleCop.Analyzers.CodeFixes.dll

Additional information: Unable to cast object of type 'Microsoft.CodeAnalysis.CSharp.Syntax.AttributeListSyntax' to type 'Microsoft.CodeAnalysis.CSharp.Syntax.AttributeSyntax'.

   at StyleCop.Analyzers.ReadabilityRules.SA1133CodeFixProvider.<GetTransformedDocumentAsync>d__5.MoveNext()
```

(*VS2015 update 1, StyleCop 1.0 RC1*)

### PR DESCRIPTION
**Pull request by NikolayIT**
Fix #1878

### CODE REVIEW

File: StyleCop.Analyzers/StyleCop.Analyzers.CodeFixes/ReadabilityRules/SA1133CodeFixProvider.cs [hunk 1]
```diff
@@ -51,8 +51,18 @@ public override Task RegisterCodeFixesAsync(CodeFixContext context)
         private static async Task<Document> GetTransformedDocumentAsync(Document document, Diagnostic diagnostic, CancellationToken cancellationToken)
         {
             var syntaxRoot = await document.GetSyntaxRootAsync(cancellationToken).ConfigureAwait(false);
-            var violatingAttribute = (AttributeSyntax)syntaxRoot.FindNode(diagnostic.Location.SourceSpan).Parent;
-            var attributeList = (AttributeListSyntax)violatingAttribute.Parent;
+            var nodeInSourceSpan = syntaxRoot.FindNode(diagnostic.Location.SourceSpan);
```
Comment (sharwell) [inline StyleCop.Analyzers/StyleCop.Analyzers.CodeFixes/ReadabilityRules/SA1133CodeFixProvider.cs]: We've run into problems in the past where error nodes in the tree caused `FindNode` to return unexpected nodes. Normally this is solved by setting `getInnermostNodeForTie: true`, but I'm not sure that would behave properly here. We probably want to have a third case handled where `nodeInSourceSpan.Parent` is neither an `AttributeListSyntax` nor an `AttributeSyntax`, and just return.

**Comment by NikolayIT** [comment]
@sharwell added default case which just returns the document object

**Comment by Noryoko** [comment]
Could you add these tests?

```
[Foo]
[Bar, Car]

[Foo, Bar]
[Car]

[Foo, Bar, Car]

[Foo]
[Bar, Car]
[Ear]
```

**Comment by NikolayIT** [comment]
@Noryoko unit tests added.

**Comment by NikolayIT** [comment]
@Noryoko, @vweijsters, do you think the following tests should also pass?

    [InlineData("[Foo][Bar, Car]", "[Foo]\r\n[Bar]\r\n[Car]", 2, 12)]
    [InlineData("[Foo, Bar][Car]", "[Foo]\r\n[Bar]\r\n[Car]", 2, 7)]
    [InlineData("[Foo][Bar, Car][Ear]", "[Foo]\r\n[Bar]\r\n[Car]\r\n[Ear]", 2, 12)]

With the current code they are failing.

File: StyleCop.Analyzers/StyleCop.Analyzers.CodeFixes/ReadabilityRules/SA1133CodeFixProvider.cs (hunk 1 above)
Comment (vweijsters) [inline StyleCop.Analyzers/StyleCop.Analyzers.CodeFixes/ReadabilityRules/SA1133CodeFixProvider.cs]: :exclamation:  A better solution would be to use `getInnermostNodeForTie: true` and then call `FirstAncestorOrSelf<AttributeListSyntax>()` on the found node. That call should never fail, as the analyzer works on `AttributeSyntaxList`.

**Comment by Noryoko** [comment]
I think so. The fix is already splitting the attributes onto their own lines. Would make the end result look consistent.

**Comment by vweijsters** [comment]
I don't think they should pass, as they are partly in the domain of SA1134. The SA1133 code fix adds the new lines to make sure that it doesn't immediately introduces a SA1134 after code fixing, but it will not / should not attempt to fix any preexisting SA1134 cases.

**Comment by NikolayIT** [comment]
OK, then this pull request is ready for approval.

**Comment by Noryoko** [comment]
I'll change my vote to agree with @vweijsters.

**Comment by vweijsters** [comment]
@NikolayIT I would like to see my remark on `SA1133CodeFixProvider.cs` to be addressed before I'm ok with this pull request.

**Comment by NikolayIT** [comment]
@vweijsters is that what you meant?

**Comment by vweijsters** [comment]
No, I meant:
> :exclamation: A better solution would be to use getInnermostNodeForTie: true and then call FirstAncestorOrSelf<AttributeListSyntax>() on the found node. That call should never fail, as the analyzer works on AttributeSyntaxList.

**Comment by NikolayIT** [comment]
Sorry, didn't see it. Is now OK? @vweijsters

**Comment by vweijsters** [comment]
Totally :wink: 

:+1: Looks good to me

**Comment by Noryoko** [comment]
:+1: Looks good to me

**Comment by vweijsters** [comment]
@sharwell Could you merge this into master as well?

**Comment by sharwell** [comment]
@vweijsters yes I'll merge it tonight

**Pull request merged**